--- QueryReadStore.js.old	2009-09-24 07:08:44.000000000 -0600
+++ QueryReadStore.js	2009-09-24 08:25:23.000000000 -0600
@@ -59,6 +59,9 @@
 		
 		url:"",
 		requestMethod:"get",
+		
+		//send these args to the xhr call every time
+		xhrArgs:null,
 		//useCache:false,
 		
 		// We use the name in the errors, once the name is fixed hardcode it, may be.
@@ -107,6 +110,7 @@
 		
 		constructor: function(/* Object */ params){
 			dojo.mixin(this,params);
+			this.xhrArgs = this.xhrArgs || {};
 		},
 		
 		getValue: function(/* item */ item, /* attribute-name-string */ attribute, /* value? */ defaultValue){
@@ -396,7 +400,7 @@
 				fetchHandler(this._items, request, this._numRows);
 			}else{
 				var xhrFunc = this.requestMethod.toLowerCase()=="post" ? dojo.xhrPost : dojo.xhrGet;
-				var xhrHandler = xhrFunc({url:this.url, handleAs:"json-comment-optional", content:serverQuery});
+				var xhrHandler = xhrFunc(dojo.mixin({url:this.url, handleAs:"json-comment-optional", content:serverQuery}, this.xhrArgs));
 				xhrHandler.addCallback(dojo.hitch(this, function(data){
 					this._xhrFetchHandler(data, request, fetchHandler, errorHandler);
 				}));
