Index: Parsing.py
===================================================================
RCS file: /home/cvs2/third_party/pyrex/Pyrex/Compiler/Parsing.py,v
retrieving revision 1.2
diff -u -r1.2 Parsing.py
--- Parsing.py	1 Jul 2006 22:24:00 -0000	1.2
+++ Parsing.py	12 Nov 2006 23:37:58 -0000
@@ -1872,7 +1872,7 @@
     while s.sy != stop:
         expr.append (s.systring)
         s.next()
-    return ''.join (expr)
+    return ' '.join (expr)
 
 #----------------------------------------------
 #
