--- Application.py	Wed Jan 22 10:03:25 2003
+++ Application.py	Sun Feb  1 18:03:03 2004
@@ -389,7 +389,10 @@
 		try:
 			ssPath = request.serverSidePath()
 			if ssPath is None or not os.path.exists(ssPath):
-				self.handleBadURL(transaction)
+				try:
+					self.handleGoodURL(transaction)
+				except Exception, exc:
+					self.handleBadURL(transaction)
 			elif isdir(ssPath) and noslash(request.pathInfo()): # (*) see below
 				self.handleDeficientDirectoryURL(transaction)
 			elif self.isSessionIdProblematic(request):
@@ -1030,7 +1033,7 @@
 			cache = self._servletCacheByPath.get(path, None)
 
 		# File is not newer?
-		if cache and cache['timestamp']<os.path.getmtime(path):
+		if cache and cache.has_key('timestamp') and cache['timestamp']<os.path.getmtime(path):
 			cache['instances'][:] = []
 			cache = None
 
@@ -1038,11 +1041,15 @@
 			cache = {
 				'instances':  [],
 				'path':	      path,
-				'timestamp':  os.path.getmtime(path),
 				'threadsafe': 0,
 				'reuseable':  0,
 				}
 
+			try:
+				cache['timestamp'] = os.path.getmtime(path)
+			except OSError:
+				pass
+
 			self.newServletCacheItem(path,cache)
 			inst = self.getServlet(transaction,path,cache)
 
@@ -1284,8 +1291,8 @@
 			else:
 				return None, None, None
 
-		elif not os.path.isfile(ssPath):
-			return None, None, None
+		#elif not os.path.isfile(ssPath):
+		#	return None, None, None
 
 		self._serverSideInfoCacheByPath[urlPath] = ssPath, contextPath, contextName
 
