--- setup.py.orig	2008-08-31 20:40:36.000000000 +0800
+++ setup.py	2008-08-31 20:46:26.000000000 +0800
@@ -6,7 +6,7 @@
 import os
 import glob
 try:
-    from distutils.core import setup
+    from setuptools import setup
     from distutils.command.build_py import build_py
     from distutils.command.install_data import install_data
 except ImportError:
@@ -75,7 +75,7 @@
     'version': '0.5',
     'author': 'David Goodger',
     'author_email': 'goodger@python.org',
-    'license': 'public domain, Python, BSD, GPL (see COPYING.txt)',
+    'license': 'public-domain PYTHON BSD GPL',
     'platforms': 'OS-independent',
     'package_dir': {'docutils': 'docutils', '': 'extras'},
     'packages': ['docutils',
