--- setup.py.orig	2008-08-31 20:03:36.000000000 +0800
+++ setup.py	2008-08-31 20:03:58.000000000 +0800
@@ -6,7 +6,7 @@
 # Usage: python setup.py install
 #
 
-from distutils.core import setup
+from setuptools import setup
 
 try:
     # add download_url syntax to distutils
@@ -32,7 +32,7 @@
     description=DESCRIPTION,
     long_description=LONG_DESCRIPTION,
     download_url="http://effbot.org/downloads#elementtree",
-    license="Python (MIT style)",
+    license="ElementTree",
     packages=["elementtree"],
     platforms="Python 1.5.2 and later.",
     classifiers=[
