--- Adapters/CGIAdapter.py	2005-08-25 01:09:02.759432184 +0200
+++ Adapters/CGIAdapter.py	2005-08-25 01:09:05.778973144 +0200
@@ -48,7 +48,7 @@
 				msvcrt.setmode(sys.stdin.fileno(), os.O_BINARY)
 
 			myInput = ''
-			if os.environ.has_key('CONTENT_LENGTH'):
+			if os.environ.has_key('CONTENT_LENGTH') and os.environ['CONTENT_LENGTH']:
 				length = int(os.environ['CONTENT_LENGTH'])
 				myInput = myInput + sys.stdin.read(length)
 
