--- HTTPRequest.py	Thu Jul 31 23:08:06 2003
+++ HTTPRequest.py	Sun Feb  1 18:03:56 2004
@@ -240,6 +240,53 @@
 		# @@ 2000-03-26 ce: maybe belongs in section below. clean up docs
 		return self._environ['REMOTE_USER']
 
+	## Languages ##
+
+	def _contentPreferences(self, accept_preference):
+		"""
+		Returns the preferences as requested by the user agent. The preferences are
+		returned as a list of codes in the same order as they appeared in the
+		appropriate environment variable. In other words, the explicit weighting
+		criteria are ignored.
+
+		As the 'accept_preference' parameter, values for language and charset
+		preferences are appropriate.
+		"""
+
+		accept_defs = accept_preference.split(",")
+		accept_prefs = []
+		for accept_def in accept_defs:
+			t = accept_def.split(";")
+			if len(t) >= 1:
+				accept_prefs.append(t[0].strip())
+		return accept_prefs
+
+	def contentLanguages(self):
+		"""
+		Returns the language preferences.
+		"""
+
+		try:
+			return self._contentPreferences(self._environ["HTTP_ACCEPT_LANGUAGE"])
+		except KeyError:
+			return []
+
+	def contentCharsets(self):
+		"""
+		Returns the character set preferences.
+		"""
+
+		try:
+			return self._contentPreferences(self._environ["HTTP_ACCEPT_CHARSET"])
+		except KeyError:
+			return []
+
+	def contentType(self):
+		"""
+		Returns the content type of the request.
+		"""
+
+		return self._environ.get('CONTENT_TYPE', None)
 
 	## Remote info ##
 
@@ -503,7 +550,7 @@
 		you want to be sure you get the entire body of the request.
 		"""
 		fs = self.fieldStorage()
-		if rewind:
+		if fs is not None and fs.file is not None and rewind:
 			fs.file.seek(0)
 		return fs.file
 
