diff -Naur /usr/lib/python2.4/site-packages/topp.build.lib-0.3.3-py2.4.egg/topp/build/lib/commands.py topp.build.lib-0.3.3-py2.4.egg/topp/build/lib/commands.py
--- /usr/lib/python2.4/site-packages/topp.build.lib-0.3.3-py2.4.egg/topp/build/lib/commands.py	2007-05-03 08:53:04.000000000 -0400
+++ topp.build.lib-0.3.3-py2.4.egg/topp/build/lib/commands.py	2007-05-22 09:39:14.000000000 -0400
@@ -16,6 +16,26 @@
 
 join = os.path.join
 
+## Added to support systems where tar is not GNU tar and
+## make is not GNU make.   -- RJK
+def getCommand( commandList, argument, findString, commandName):
+    for n in commandList:
+        out = subprocess.Popen("%s %s"%(n, argument),  shell=True, bufsize=512, stdout=subprocess.PIPE).stdout.read()
+        if out.find(findString)!=-1:
+            return n 
+    raise Exception("Unable to find %s"%commandName)
+
+def getGNUTarCommand():
+    #common GNU tar names
+    tarNames = ["tar", "gtar"]
+    return getCommand(tarNames, "--version", "GNU", "GNU Version of 'tar'")
+        
+        
+def getGNUMakeCommand():
+    #comman GNU make names
+    makeNames = ["make", "gmake"]
+    return getCommand(makeNames, "--version", "GNU", "GNU Version of 'make'")
+
 class WorkingEnv(ShellCommand):
     """ make and activate a workingenv """
 
@@ -102,7 +122,7 @@
                                   stdout=subprocess.PIPE)
             stdin = cat.stdout
         
-        cmdline = ['tar', 'xv']
+        cmdline = [ getGNUTarCommand() , 'xv']
         if self.path:
             cmdline += ['-C', self.path]
 
diff -Naur /usr/lib/python2.4/site-packages/topp.build.lib-0.3.3-py2.4.egg/topp/build/lib/__init__.py topp.build.lib-0.3.3-py2.4.egg/topp/build/lib/__init__.py
--- /usr/lib/python2.4/site-packages/topp.build.lib-0.3.3-py2.4.egg/topp/build/lib/__init__.py	2007-05-03 08:53:04.000000000 -0400
+++ topp.build.lib-0.3.3-py2.4.egg/topp/build/lib/__init__.py	2007-05-22 09:53:31.000000000 -0400
@@ -1058,3 +1058,9 @@
         for i in self.apps:
             self.builder[i].perform_tasks()
         self.perform_tasks() 
+
+
+
+
+    
+


!DSPAM:4014,4652f8eb20994901796417!
