diff -Naur /usr/lib/python2.4/site-packages/topp.build.opencore-0.1.4dev_r4918-py2.4.egg/topp/build/opencore/tasks/zope.py topp.build.opencore-0.1.4dev_r4918-py2.4.egg/topp/build/opencore/tasks/zope.py
--- /usr/lib/python2.4/site-packages/topp.build.opencore-0.1.4dev_r4918-py2.4.egg/topp/build/opencore/tasks/zope.py	2007-05-14 08:22:03.000000000 -0400
+++ topp.build.opencore-0.1.4dev_r4918-py2.4.egg/topp/build/opencore/tasks/zope.py	2007-05-22 09:41:06.000000000 -0400
@@ -6,6 +6,7 @@
 
 from buildit.task import Task
 from topp.build.lib.taskfuncs import copyskeleton
+from topp.build.lib.commands import getGNUMakeCommand
 from buildit.commandlib import Patch
 
 checkout = Task(
@@ -41,8 +42,8 @@
     targets=['${zope_source}/bin',],
     workdir='${zope_source}',
     commands=['./configure --prefix=${zope_home}',
-              'make',
-              'make inplace',
+              '%s'%getGNUMakeCommand(),
+              '%s inplace'%getGNUMakeCommand(),
               ],
     dependencies=(patch_zope,)
     )


!DSPAM:4014,4652f8eb20994901796417!
