\chapter{Introduction}
\label{cha:introduction}

\begin{quote}
This chapter introduces the \PyPedal{} module for Python 2.4 and outlines the rest of the document.
\end{quote}

\PYPEDAL{} (\textbf{P}ython \textbf{Ped}igree An\textbf{al}ysis) is a tool for analyzing pedigree files.  It calculates several quantitative measures of allelic and genoytpic diversity from pedigrees, including average coefficients of inbreeding and relationship, effective number of founders, and effective number of ancestors.  Some qualitative checks are performed in order to catch some common mistakes, such as parents with more recent birthdates or ID numbers than their offspring.  Tools for pedigree visualization and report generation are provided.  Currently, \PYPEDAL{} only makes use of information on pedigree structure.  Allelotypes can be assigned to founders (or read from the pedigree file) for use in gene-dropping simulations to compute effective number of  founder genomes, but no other measures of alleic diversity are currently supported.

\PYPEDAL{} is a Python (\url(http://www.python.org/))language module that may be called by other Python programs or used interactively from the Python interpreter.  You must have Python 2.4 installed in order to use \PYPEDAL() as \PYPEDAL() makes use of some version-specific features found only in 2.4.  The Numarray module must be installed in order for you to use \PYPEDAL(), and may be found at \url{http://www.stsci.edu/resources/software_hardware/numarray}.

This document is the ``official'' documentation for \PyPedal{}. It includes a tutorial and is the most authoritative source of information about \PyPedal{} with the exception of the source code. The tutorial material will walk you through a set of manipulations of a simple pedigree.  All users of \PYPEDAL{} are encouraged to follow the tutorial with a working \PYPEDAL{} installation. The best way to learn is by doing --- the aim of this tutorial is to guide you along this "doing."

This content of this manual is broken down as follows:
\begin{description}
\item[License] Chapter \ref{cha:license} describes the license under which \PyPedal{} is distributed.  It is important that you review the license before using the program.
\item[Installing PyPedal] Chapter \ref{cha:installation} provides information
   on testing Python and installing PyPedal.
\item[High-Level Overview] Chapter \ref{cha:high-level-overview} gives a
   high-level overview of the components of the \PyPedal{} system as a whole.
\item[Applications Programming Interface] Chapter \ref{cha:api} includes a complete reference, including useage notes, for all functions in all \PyPedal{}. modules.
\item[PyPedal Tutorial] Chapter \ref{cha:tutorial} provides a gentle introduction to \PyPedal{}.
\item[Glossary] Chapter \ref{cha:glossary} provides a glossary of terms.
\item[References and Indices] are provided at the end of the manual.
\end{description}

\section{Implemented Features}
PyPedal is currently capable of doing the following things:
\begin{itemize}
\item Reading pedigree files in user-defined formats;
\item Checking pedigree integrity (duplicate IDs, parents younger than offspring, etc.);
\item Generating summary information such as frequency of appearance in the pedigree file;
\item Computation of the numerator relationship matrix ($A$) from a pedigree file using the tabular method;
\item Inbreeding calculations for large pedigrees;
\item Computation of average total and average individual coefficients of inbreeding and relationship;
\item Decomposition of $A$ into $T$ and $D$ such that $A=TDT'$;
\item Computation of the direct inverse of $A$ (not accounting for inbreeding) using the method of Henderson \cite{ref143};
\item Computation of the direct inverse of $A$ (accounting for inbreeding) using the method of \citeN{ref235};
\item Storage of $A$ and its inverse between user sessions as persistent Python objects using the pickle module to avoid unnecessary calculations;
\item Calculation of theoretical effective population size;
\item Calculation of actual effective population size based on the change in population average inbreeding;
\item Computation of effective founder number using the exact algorithm of \citeN{ref640};
\item Computation of effective founder number using the approximate algorithm of \citeN{ref352};
\item Computation of effective ancestor number using the algorithm of \citeN{ref352};
\item Selection of subpedigrees containing all ancestors of an animal;
\item Identification of the common relatives of two animals;
\item Output to ASCII text files, including matrices, coefficients of inbreeding and relationship, and summary information;
\item Reordering and renumbering of pedigree files.
\end{itemize}
A full list of features, including notes on useage and computational details, is provided in Chapter \ref{cha:api}.  \PYPEDAL{} has been used to perform calculations on pedigrees as large as 100,000 animals and has used in scientific research \cite{Cole2004a}.

\section{Where to get information and code}
\PYPEDAL{} and its documentation are available at: \url{http://pypedal.sourceforge.net/}. The Numarray web site is: \url{http://numpy.sourceforge.net/}. The Python web site is \url{http://www.python.org/}.

\section{Acknowledgments}
\PYPEDAL{} was initially written to support the author's dissertation research while at Louisiana State University, Baton Rouge (\url{http://www.lsu.edu/}).  It lay fallow for some time but has recently come under active development again.  This is due in part to a request from colleagues at the University of Minnesota that led to the inclusion of new functionality in \PYPEDAL{}.  The author wishes to thank Dr. Paul Van{R}aden for very helpful suggestions for improving the ability of \PYPEDAL{} to handle certain computations in very large pedigrees.  Additional feedback in the form of bug reports, feature requests, and discussion of computing strategies was provided by Edward H. Hagen (Institute for Theoretical Biology, Humboldt-Universitï¿½zu Berlin), Kathy Hanford (University of Nebraska, Lincoln), Thomas von Hassell, and Gianluca Saba.