--- /usr/lib/python2.3/tarfile.py	2004-05-06 11:47:11.306150232 +0200
+++ itarfile.py	2004-05-06 11:54:14.000000000 +0200
@@ -1,7 +1,15 @@
 #!/usr/bin/env python
 # -*- coding: iso-8859-1 -*-
 #-------------------------------------------------------------------
-# tarfile.py
+# itarfile.py
+#-------------------------------------------------------------------
+# This is a patched version of python2.3.3 orginal tarfile to not
+# set the size on folderish entries inside the tarfile which is not
+# supported by tons of tarreaders out there.
+#
+# contact Simon Eisenmann <simon@struktur.de> 
+# for infos on the patch.
+#
 #-------------------------------------------------------------------
 # Copyright (C) 2002 Lars Gustäbel <lars@gustaebel.de>
 # All rights reserved.
@@ -1120,7 +1128,7 @@
         tarinfo.mode  = stmd
         tarinfo.uid   = statres.st_uid
         tarinfo.gid   = statres.st_gid
-        tarinfo.size  = statres.st_size
+        tarinfo.size  = not stat.S_ISDIR(stmd) and statres.st_size or 0
         tarinfo.mtime = statres.st_mtime
         tarinfo.type  = type
         tarinfo.linkname = linkname
